Enable HSTS with IncludeSubdomains header by default for new apps

prathamesh-sonpatki opened the pull request:
Fixes https://github.com/rails/rails/pull/22663

homakov committed:
HSTS without IncludeSubdomains is often useless

1) Because if you forget to add Secure; to the session cookie, it will leak to http:// subdomain in some cases
2) Because http:// subdomain can Cookie Bomb/cookie force main domain or be used for phishing.

That's why *by default* it must include subdomains as it's much more common scenario. Very few websites *intend* to leave their blog.app.com working over http:// while having everything else encrypted. 

Yes, many developers forget to add subdomains=true by default, believe me :)

rafaelfranca commented on railties/lib/rails/generators/rails/app/templates/config/initializers/hsts_subdomains.rb:
```diff
@@ -0,0 +1 @@
+Rails.application.config.action_dispatch.hsts_subdomains = true
```
Instead of a new option can we set `Rails.application.config.action_dispatch.ssl_options = { subdomains: true }`?

prathamesh-sonpatki replied:
@rafaelfranca That will mess up with existing ssl_options present in `config/environments/production.rb` right?

rafaelfranca replied:
Is there any existent in production.rb? If that is the case we can try to use `merge` with the already existent value.

prathamesh-sonpatki replied:
Ah okey `config.ssl_options` is different that `config.action_dispatch.ssl_options`. 

But you meant reusing `config.ssl_options` right?

rafaelfranca replied:
Right. `config.ssl_options`

prathamesh-sonpatki replied:
@rafaelfranca We can do 

``` ruby
Rails.application.config.ssl_options.merge!({ hsts: { subdomains: true } })
```

But that can mess up with setting the default hsts options here - https://github.com/prathamesh-sonpatki/rails/blob/master/actionpack/lib/action_dispatch/middleware/ssl.rb#L82.

Note that we are passing options only in case of https://github.com/prathamesh-sonpatki/rails/blob/master/actionpack/lib/action_dispatch/middleware/ssl.rb#L92. 

options are `config.ssl_options[:hsts]` 

So if we want to reuse config.ssl_options then we will have to update each branch in https://github.com/prathamesh-sonpatki/rails/blob/master/actionpack/lib/action_dispatch/middleware/ssl.rb#L82 to pass options[:subdomains].

It will override user's config.ssl_options[:hsts] though as we will have a default hash instead of true/false/nil/custom options.

rafaelfranca replied:
I don't get that is the problem. If we merged `{ expires: HSTS_EXPIRES_IN, subdomains: false, preload: false }` with `{ subdomain: true }` would not the result be `{ expires: HSTS_EXPIRES_IN, subdomains: true, preload: false }`?

```irb
>> default = { expires: 1, subdomains: false, preload: false }
=> {:expires=>1, :subdomains=>false, :preload=>false}
>> default.merge(subdomains: true)
=> {:expires=>1, :subdomains=>true, :preload=>false}
```

prathamesh-sonpatki commented on railties/lib/rails/generators/rails/app/templates/config/initializers/hsts_subdomains.rb:
```diff
@@ -0,0 +1 @@
+Rails.application.config.ssl_options.reverse_merge!({ hsts: { subdomains: true } })
```
@rafaelfranca Made the change as per our discussion. This will make sure that we don't mess up with user's `config.ssl_options[:hsts]`. It can be false or true or whatever, it will still remain that value after this initializer will get run.

rafaelfranca replied:
We can just set the ssl_option directly, not need to merge. This line will be generated only in new applications, so there is no `ssl_options` set anywhere. For existent applications we will let the deprecation warning handle it.

We will need to update the `config_when_updating` method in the generators to not generate this file.

Also it should be named `ssl_options.rb`

prathamesh-sonpatki commented on actionpack/lib/action_dispatch/middleware/ssl.rb:
```diff
@@ -40,7 +40,7 @@ class SSL
     def self.default_hsts_options
-      { expires: HSTS_EXPIRES_IN, subdomains: false, preload: false }
+      { expires: HSTS_EXPIRES_IN, subdomains: true, preload: false }
     end
```
This is still shaky, thinking about how to handle it properly.

rafaelfranca replied:
The default should be `false`

prathamesh-sonpatki committed:
New applications will be generated with ssl_options to enable HSTS with subdomains

- We will reuse config.ssl_options for setting the HSTS settings.

prathamesh-sonpatki committed:
Old applications will not get the ssl_options initializer

- We will remove the initializer for old apps which are migrated to
  Rails 5 so that they are not affected by this breaking change.

prathamesh-sonpatki committed:
Added deprecation for older apps

- For old apps which are not setting any value for hsts[:subdomains],
  a deprecation warning will be shown saying that hsts[:subdomains] will
  be turned on by default in Rails 5.1. Currently it will be set to
  false for backward compatibility.
- Adjusted tests to reflect this change.

prathamesh-sonpatki committed:
Add CHANGELOG entry

prathamesh-sonpatki committed:
Update documentation and deprecation message

prathamesh-sonpatki commented:
r? @rafaelfranca Please review.
